network/libp2p_impl.go

```diff
@@ -160,19 +163,36 @@ func (dtnet *libp2pDataTransferNetwork) SendMessage(
 	p peer.ID,
 	outgoing datatransfer.Message) error {
 
+	ctx, span := otel.Tracer("data-transfer").Start(ctx, "sendMessage", trace.WithAttributes(
+		attribute.String("to", p.String()),
+		attribute.Int64("transferID", int64(outgoing.TransferID())),
+		attribute.Bool("isRequest", outgoing.IsRequest()),
+		attribute.Bool("isNew", outgoing.IsNew()),
+		attribute.Bool("isRestart", outgoing.IsRestart()),
+		attribute.Bool("isUpdate", outgoing.IsUpdate()),
+		attribute.Bool("isCancel", outgoing.IsCancel()),
+		attribute.Bool("isPaused", outgoing.IsPaused()),
+	))
+
+	defer span.End()
 	s, err := dtnet.openStream(ctx, p, dtnet.dtProtocols...)
 	if err != nil {
+		span.RecordError(err)
 		return err
 	}
 
 	outgoing, err = outgoing.MessageForProtocol(s.Protocol())
 	if err != nil {
-		return xerrors.Errorf("failed to convert message for protocol: %w", err)
+		err = xerrors.Errorf("failed to convert message for protocol: %w", err)
+		span.RecordError(err)
+		return err
 	}
 
 	if err = dtnet.msgToStream(ctx, s, outgoing); err != nil {
+		span.RecordError(err)
 		if err2 := s.Reset(); err2 != nil {
 			log.Error(err)
+			span.RecordError(err2)
 			return err2
 		}
 		return err
```

Comment:
```suggestion
		if err2 := s.Reset(); err2 != nil {
			log.Error(err)
			span.RecordError(err2)
			span.SetStatus(codes.Error, err2.Error())
			return err2
		}
		span.RecordError(err)
		span.SetStatus(codes.Error, err.Error())
		return err
```

Reply:
not sure about that shuffle tbh, maybe recording both errors is the right approach, but the `SetStatus()` should probably record just the fatal one for each branch

Reply:
genuinely, I'm not sure what the right path is!